### PR TITLE
[multi-ASIC] BGP internal neighbor table support

### PR DESCRIPTION
**Pull request by judyjoseph**
**- Why I did it**

Introduce a new table for BGP internal neighbor sessions which will be used in case of multi-asic platforms for the IBGP sessions between the BGP instances running in namespaces mapped to ASIC. This is the follow up of the comments added for a fix earlier to create internal BGP peer group ( https://github.com/Azure/sonic-buildimage/pull/4620#issuecomment-631664961 )

**- How I did it**
Updated the minigraph parsing logic to get the bgp_internal_sessions 

**- How to verify it**
Checked the DB to find that the BGP_INTERNAL_NEIGHBOR table is populated with internal BGP neigibors.

**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [x] 201911
- [x] 202006

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: src/sonic-bgpcfgd/bgpcfgd [hunk 1]
```diff
@@ -876,6 +876,7 @@ def main():
         ZebraSetSrc(common_objs, "STATE_DB", swsscommon.STATE_INTERFACE_TABLE_NAME),
         # Peer Managers
         BGPPeerMgrBase(common_objs, "CONFIG_DB", swsscommon.CFG_BGP_NEIGHBOR_TABLE_NAME, "general", True),
+        BGPPeerMgrBase(common_objs, "CONFIG_DB", swsscommon.CFG_BGP_INTERNAL_NEIGHBOR_TABLE_NAME, "internal", True),
```
Comment (pavel-shirshov) [inline src/sonic-bgpcfgd/bgpcfgd:879]: Please put False as a last argument

File: dockers/docker-fpm-frr/frr/bgpd/templates/internal/policies.conf.j2 [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+route-map FROM_BGP_INTERNAL_PEER_V6 permit 100
+!
+route-map TO_BGP_INTERNAL_PEER_V6 permit 100
+!
+{% if CONFIG_DB__DEVICE_METADATA['localhost']['sub_role'] == 'BackEnd' %}
+route-map FROM_BGP_INTERNAL_PEER_V4 permit 2
+ set originator-id {{ loopback0_ipv4 | ip }}
+!
+route-map FROM_BGP_INTERNAL_PEER_V6 permit 1
+ set ipv6 next-hop prefer-global
```
Comment (pavel-shirshov) [inline dockers/docker-fpm-frr/frr/bgpd/templates/internal/policies.conf.j2:23]: I think we need this rule on frontend peer also

**Review by pavel-shirshov — CHANGES_REQUESTED** [review]
Can you please add tests for the new templates and update tests for the old?

**Comment by pavel-shirshov** [comment]
Can you please add template tests data in src/sonic-bgpcfgd/tests/data ?

**Comment by judyjoseph** [comment]
> Can you please add template tests data in src/sonic-bgpcfgd/tests/data ?

@pavel-shirshov I updated/added tests here in my earlier commit https://github.com/Azure/sonic-buildimage/pull/5520/commits/2b15a57b473c9c7a330ebdbb5bbb96fc408d52f7. But let me check if I need to cover more scenarios

**Comment by judyjoseph** [comment]
retest this please

File: src/sonic-bgpcfgd/tests/data/general/instance.conf/param_all_v4.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
     "CONFIG_DB__DEVICE_METADATA": {
         "localhost": {
-            "sub_role": "BackEnd"
+            "sub_role": "FrontEnd"
```
Comment (pavel-shirshov) [inline src/sonic-bgpcfgd/tests/data/general/instance.conf/param_all_v4.json:4]: I think the sub_role was removed completely from the instance.conf? We don't need it anymore

File: src/sonic-bgpcfgd/tests/data/general/instance.conf/param_all_v6.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
     "CONFIG_DB__DEVICE_METADATA": {
         "localhost": {
-            "sub_role": "BackEnd"
+            "sub_role": "FrontEnd"
```
Comment (pavel-shirshov) [inline src/sonic-bgpcfgd/tests/data/general/instance.conf/param_all_v6.json:4]: The same as previous

File: src/sonic-bgpcfgd/tests/data/general/peer-group.conf/param_base.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
     "CONFIG_DB__DEVICE_METADATA": {
         "localhost": {
             "type": "LeafRouter",
-            "sub_role": "BackEnd"
+            "sub_role": "FrontEnd"
```
Comment (pavel-shirshov) [inline src/sonic-bgpcfgd/tests/data/general/peer-group.conf/param_base.json:5]: Do we still have this?

File: src/sonic-bgpcfgd/tests/data/general/policies.conf/param_all.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
     "CONFIG_DB__DEVICE_METADATA": {
         "localhost": {
-            "sub_role": "BackEnd"
+            "sub_role": "FrontEnd"
```
Comment (pavel-shirshov) [inline src/sonic-bgpcfgd/tests/data/general/policies.conf/param_all.json:4]: The same as previous. I'd remove it, unless we need it for the template

File: src/sonic-bgpcfgd/tests/data/general/policies.conf/param_deny.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
     "CONFIG_DB__DEVICE_METADATA": {
         "localhost": {
-            "sub_role": "BackEnd"
+            "sub_role": "FrontEnd"
```
Comment (pavel-shirshov) [inline src/sonic-bgpcfgd/tests/data/general/policies.conf/param_deny.json:4]: The same as before

**Review by pavel-shirshov — APPROVED** [review]
I'd better remove not required entries.

**Comment by judyjoseph** [comment]
retest this please

**Comment by judyjoseph** [comment]
retest vsimage please

**Review by pavel-shirshov — APPROVED** [review]
LGTM

**Comment by judyjoseph** [comment]
@arlakshm , Could you take a look as well ?

File: dockers/docker-fpm-frr/frr/bgpd/templates/internal/instance.conf.j2 [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+!
+  neighbor {{ neighbor_addr }} remote-as {{ bgp_session['asn'] }}
+  neighbor {{ neighbor_addr }} description {{ bgp_session['name'] }}
+!
+{% if neighbor_addr | ipv4 %}
+  address-family ipv4
+    neighbor {{ neighbor_addr }} peer-group INTERNAL_PEER_V4
+!
+{%   if CONFIG_DB__DEVICE_METADATA['localhost']['sub_role'] == 'BackEnd' %}
+    neighbor {{ neighbor_addr }} route-map FROM_BGP_INTERNAL_PEER_V4 in
```
Comment (arlakshm) [inline dockers/docker-fpm-frr/frr/bgpd/templates/internal/instance.conf.j2:12]: Is route-map configuration line repeated?  The same line present in the peer_group.conf.j2 configuration

**Review by arlakshm — APPROVED** [review]
lgtm, couple of minor comments

File: dockers/docker-fpm-frr/frr/bgpd/templates/internal/instance.conf.j2 (hunk 8 above)
Comment (judyjoseph) [reply dockers/docker-fpm-frr/frr/bgpd/templates/internal/instance.conf.j2:12]: I looked at both .. they are not same.
Here we set it per neighbor_ip_address, while in the peer_group.conf.j2 it is set on the peer group -- similar to how it was done before as well.

File: dockers/docker-fpm-frr/frr/bgpd/templates/internal/instance.conf.j2 [hunk 9]
```diff
@@ -0,0 +1,34 @@
+!
+! template: bgpd/templates/internal/instance.conf.j2
+!
+  neighbor {{ neighbor_addr }} remote-as {{ bgp_session['asn'] }}
+  neighbor {{ neighbor_addr }} description {{ bgp_session['name'] }}
+!
```
[the remark this answers is not in the document]
Comment (judyjoseph) [reply dockers/docker-fpm-frr/frr/bgpd/templates/internal/instance.conf.j2:6]: Sure will update.

**Pull request merged**